experimental/cache_line_analysis.cc

```diff
@@ -0,0 +1,95 @@
+#include <stdlib.h>
+#include <time.h>
+
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <random>
+#include <string>
+
+// TODO: give compilation instructions instead of including nonlocal files
+#include "../demos/asm/measurereadlatency.h"
+#include "../demos/instr.h"
+#include "../demos/utils.h"
+
+// Determines the average latency of reading a subset of elements of a vector
+// "buf" from memory to cache
+// It reads elements that are located "step" size from each other (i.e. 0, step,
+// 2*step, 3* step, ...) in a random order and measures the time using
+// MeasureReadLatency(). Later this number is used to determine cache line sizes
+double_t FindReadingTime(std::vector<char> buf, int64_t sz, int step) {
+  int64_t accesses_size = sz / step;
+  std::vector<int64_t> accesses(accesses_size);
+  for (int64_t i = 0; i < sz; i += step) {
+    accesses.push_back(i);
+  }
+  std::random_device rd;
+  std::mt19937 f(rd());
+  std::shuffle(accesses.begin(), accesses.end(), f);
+
+  uint64_t total_read_latency = 0;
+
+  for (int64_t i : accesses) {
+    total_read_latency += MeasureReadLatency(&buf[i]);
+  }
+
+  return total_read_latency / accesses_size;
+}
+
+// Analyzes a range of numbers to find the size of strides that is bigger than
+// the cache line size, so not all cache lines are needed to be read from memory
+// to cache
+// Note that (1) the time of reading from cache is negiligble compared to
+// reading from memory, and (2) CPUs read cache lines from memory instead byte
+// by byte.
```

Comment:
Slightly unclear, maybe (2) should be phrased as "`(2) CPUs read memory in blocks of entire cache lines, rather than reading individual bytes`"